plugins/ccgx/fu-ccgx-pure-hid-device.c

```diff
@@ -0,0 +1,538 @@
+/*
+ * Copyright (C) 2023 Framework Computer Inc
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <string.h>
+
+#include "fu-ccgx-common.h"
+#include "fu-ccgx-firmware.h"
+#include "fu-ccgx-hpi-common.h"
+#include "fu-ccgx-pure-hid-device.h"
+
+#define ROW_SIZE 128
+#define FW1_START 0x0030
+#define FW2_START 0x0200
+#define FW1_METADATA 0x03FF
+#define FW2_METADATA 0x03FE
+#define WRITE_HEADER_SIZE 4
+
+/**
+ * FU_CCGX_PURE_HID_DEVICE_IS_IN_RESTART:
+ *
+ * Device is in restart and should not be closed manually.
+ *
+ * Since: 1.7.0
+ */
+#define FU_CCGX_PURE_HID_DEVICE_IS_IN_RESTART (1 << 0)
+
+struct _FuCcgxPureHidDevice {
+	FuHidDevice parent_instance;
+	FuCcgxFwMode operating_mode;
+	guint32 versions[FU_CCGX_FW_MODE_LAST];
+	//guint8 bootloader_info;
+	//guint8 bootmode_reason;
+	guint32 silicon_id;
+	//guint32 image_1_row;
+	//guint32 image_2_row;
+	//device_uid: // [u8; 6],
+	guint32 flash_row_size;
+	guint32 flash_size;
+};
+G_DEFINE_TYPE(FuCcgxPureHidDevice, fu_ccgx_pure_hid_device, FU_TYPE_HID_DEVICE)
+
+
+typedef struct __attribute__((packed)) {
```

Comment:
this looks perfect for the rustgen stuff in https://github.com/fwupd/fwupd/blob/main/docs/tutorial.md#the-rustgen-helper -- it'll make the code a lot simpler too.

Reply:
That looks awesome, I'll try that.

Reply:
I've pushed this to wip/framework-hdmi-dp -- feel free to push there yourself too -- I'll add you on the commit ACL now.